kotlin-codegen/runtime/src/main/java/com/squareup/moshi/MoshiSerializable.kt

```diff
@@ -0,0 +1,116 @@
+/*
+ * Copyright (C) 2018 Square, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.squareup.moshi
+
+import java.lang.ref.WeakReference
+import java.lang.reflect.Constructor
+import java.lang.reflect.InvocationTargetException
+import java.lang.reflect.ParameterizedType
+import java.lang.reflect.Type
+import java.util.Collections
+import java.util.LinkedHashMap
+import kotlin.annotation.AnnotationRetention.RUNTIME
+import kotlin.annotation.AnnotationTarget.CLASS
+
+@Retention(RUNTIME)
+@Target(CLASS)
+annotation class MoshiSerializable
```

Comment:
Should we fold the small runtime component into Moshi itself? Maybe even install the `MoshiSerializableFactory` by default? It’d need to be rewritten in .java, but that’s not a big deal.

Reply:
... I’m imagining like `@JsonObject` or something to sit alongside `@Json` as part of the core API.

Reply:
Can definitely do that, but worried people might use it and not realize they need to wire in the compiler (especially since it would only fail at runtime). What do you think?

Reply:
That's a problem regardless. We can fail at runtime with a nice message if they mess up.